domain/secret/state/types.go

```diff
@@ -247,6 +253,48 @@ func (rows secrets) toSecretRevisionRef(refs secretValueRefs) ([]*coresecrets.Se
 	return result, nil
 }
 
+type (
+	secretIDs               []secretID
+	secretExternalRevisions []secretExternalRevision
+)
+
+func (rows secretIDs) toSecretMetadataForDrain(revRows secretExternalRevisions) ([]*coresecrets.SecretMetadataForDrain, error) {
+	if len(rows) != len(revRows) {
+		// Should never happen.
+		return nil, errors.New("row length mismatch composing secret results")
+	}
+
+	var (
+		result  []*coresecrets.SecretMetadataForDrain
+		current *coresecrets.SecretMetadataForDrain
+	)
+	for i, row := range rows {
+		if current == nil || current.URI.ID != row.ID {
+			// Encountered a new record.
+			uri, err := coresecrets.ParseURI(row.ID)
+			if err != nil {
+				return nil, errors.NotValidf("secret URI %q", row.ID)
+			}
+			md := coresecrets.SecretMetadataForDrain{
+				URI: uri,
+			}
+			current = &md
+			result = append(result, &md)
```

Comment:
result = append(result, &current)

Reply:
either way works. if we use current it should be `append(result, current)` since current is a pointer